src/huggingface_hub/hf_api.py

```diff
@@ -1654,7 +1794,22 @@ def list_datasets(
             if isinstance(filter, DatasetFilter):
                 params = self._unpack_dataset_filter(filter)
             else:
-                params.update({"filter": filter})
+                params = self._build_dataset_filter(
+                    author=author,
+                    benchmark=benchmark,
+                    dataset_name=dataset_name,
+                    language_creators=language_creators,
+                    language=language,
+                    multilinguality=multilinguality,
+                    size_categories=size_categories,
+                    task_categories=task_categories,
+                    task_ids=task_ids,
+                )
+
+                if isinstance(filter, str):
+                    params.update({"filter": params.get("filter", set()).add(filter)})
+                else:
+                    params.update({"filter": params.get("filter", set()).update(filter)})
```

Comment:
Used set to avoid duplication in cases where a user passes both keyword argument and a filter
e.g.
```python

filter="language:en"
language="language:en"
```
can revert it to a tuple if needed

Reply:
_I think_ it's fine to keep it simple without the `set` overhead. I don't think we should optimize for users passing `list_datasets(..., filter="language:en", language="language:en")` (which is very unlikely to happen). In any case, if we the parameter is duplicated in the URL, the search query should still work so no need to worry about it.

---
(nit) to handle both raw strings and lists in such a case I would do something like this:
```py
                if isinstance(filter, str):
                    filter = [filter]
                params.update({"filter": params.get("filter", set()).update(filter)})
```
which is clearer to read IMO. Drawback is that the `filter` value is modified but that's not really an issue in this precise case.